File: src/sdk/go/k1s0/test-fixtures/mock_builder.go
// src/sdk/go/k1s0/test-fixtures/mock_builder.go
//
// k1s0 Go SDK test-fixtures: Mock builder fluent API（領域 3、ADR-TEST-010 §3）。
// リリース時点で 3 service（State / Audit / PubSub）の builder を提供。
// 採用初期で +3 (Workflow / Decision / Secret)、運用拡大時で残 6 service を追加する。
package testfixtures

// MockBuilderRoot は 12 service の mock builder への entry point
type MockBuilderRoot struct {
	// defaultTenant は WithTenant 未指定時の既定 tenant
	defaultTenant string
}

// newMockBuilderRoot は Setup 内から呼ばれる constructor
func newMockBuilderRoot(defaultTenant string) *MockBuilderRoot {
	return &MockBuilderRoot{defaultTenant: defaultTenant}
}

// State は State service の mock builder を返す（fluent chain entry）
func (r *MockBuilderRoot) State() *StateMockBuilder {
	return &StateMockBuilder{tenant: r.defaultTenant}
}

// Audit は Audit service の mock builder を返す
func (r *MockBuilderRoot) Audit() *AuditMockBuilder {
	return &AuditMockBuilder{tenant: r.defaultTenant}
}

// PubSub は PubSub service の mock builder を返す
func (r *MockBuilderRoot) PubSub() *PubSubMockBuilder {
	return &PubSubMockBuilder{tenant: r.defaultTenant}
}

// Workflow は採用初期で実装する Workflow service mock builder（panic で警告）
func (r *MockBuilderRoot) Workflow() *workflowMockBuilder {
	panic("ADR-TEST-010 PHASE: Workflow mock builder は採用初期で実装、リリース時点未対応")
}

// Decision は採用初期で実装
func (r *MockBuilderRoot) Decision() *decisionMockBuilder {
	panic("ADR-TEST-010 PHASE: Decision mock builder は採用初期で実装、リリース時点未対応")
}

// Secret は採用初期で実装
func (r *MockBuilderRoot) Secret() *secretMockBuilder {
	panic("ADR-TEST-010 PHASE: Secret mock builder は採用初期で実装、リリース時点未対応")
}

// Pii / Feature / Telemetry / Log / Binding / Invoke は運用拡大時で実装（同パタン）
func (r *MockBuilderRoot) Pii() *unimplementedMockBuilder {
	panic("ADR-TEST-010 PHASE: Pii mock builder は運用拡大時で実装")
}

// State / Audit / PubSub の builder fluent chain ----------------------

// StateMockBuilder は State service mock data の fluent builder
type StateMockBuilder struct {
	tenant string
	key    string
	value  []byte
	ttl    int
}

// WithTenant は tenant 指定で builder を更新（fluent chain）
func (b *StateMockBuilder) WithTenant(tenant string) *StateMockBuilder {
	b.tenant = tenant
	return b
}

// WithKey は key 指定
func (b *StateMockBuilder) WithKey(key string) *StateMockBuilder {
	b.key = key
	return b
}

// WithValue は value 指定（byte slice）
func (b *StateMockBuilder) WithValue(value []byte) *StateMockBuilder {
	b.value = value
	return b
}

// WithTTL は TTL 秒指定
func (b *StateMockBuilder) WithTTL(seconds int) *StateMockBuilder {
	b.ttl = seconds
	return b
}

// Build は最終的な StateEntry struct を返す（採用初期で contracts/proto 生成型に置換）
func (b *StateMockBuilder) Build() *StateEntry {
	return &StateEntry{
		Tenant: b.tenant,
		Key:    b.key,
		Value:  b.value,
		TTL:    b.ttl,
	}
}

// StateEntry は State service の wire 形式（採用初期で contracts/proto 型に置換）
type StateEntry struct {
	Tenant string
	Key    string
	Value  []byte
	TTL    int
}

// AuditMockBuilder は Audit service mock data の fluent builder
type AuditMockBuilder struct {
	tenant     string
	entryCount int
	startSeq   int
}

// WithTenant は tenant 指定
func (b *AuditMockBuilder) WithTenant(tenant string) *AuditMockBuilder {
	b.tenant = tenant
	return b
}

// WithEntries は entry 件数指定（hash chain で連結された N 件を生成）
func (b *AuditMockBuilder) WithEntries(n int) *AuditMockBuilder {
	b.entryCount = n
	return b
}

// WithSequence は開始 sequence 番号指定
func (b *AuditMockBuilder) WithSequence(seq int) *AuditMockBuilder {
	b.startSeq = seq
	return b
}

// Build は AuditEntry の slice を返す（採用初期で hash chain 計算 + contracts/proto 型に置換）
func (b *AuditMockBuilder) Build() []*AuditEntry {
	entries := make([]*AuditEntry, b.entryCount)
	for i := 0; i < b.entryCount; i++ {
		entries[i] = &AuditEntry{
			Tenant:   b.tenant,
			Sequence: b.startSeq + i,
			// 採用初期で prev_id chain を SHA-256 で計算する
			PrevID: "",
		}
	}
	return entries
}

// AuditEntry は Audit service の wire 形式（採用初期で contracts/proto 型に置換）
type AuditEntry struct {
	Tenant   string
	Sequence int
	PrevID   string
}

// PubSubMockBuilder は PubSub service mock data の fluent builder
type PubSubMockBuilder struct {
	tenant      string
	topic       string
	messages    int
	delayMs     int
}

// WithTenant は tenant 指定
func (b *PubSubMockBuilder) WithTenant(tenant string) *PubSubMockBuilder {
	b.tenant = tenant
	return b
}

// WithTopic は topic 指定
func (b *PubSubMockBuilder) WithTopic(topic string) *PubSubMockBuilder {
	b.topic = topic
	return b
}

// WithMessages は publish する message 数
func (b *PubSubMockBuilder) WithMessages(n int) *PubSubMockBuilder {
	b.messages = n
	return b
}

// WithDelayMs は publish 間隔（ミリ秒）
func (b *PubSubMockBuilder) WithDelayMs(ms int) *PubSubMockBuilder {
	b.delayMs = ms
	return b
}

// Build は PubSub message の slice を返す
func (b *PubSubMockBuilder) Build() []*PubSubMessage {
	msgs := make([]*PubSubMessage, b.messages)
	for i := 0; i < b.messages; i++ {
		msgs[i] = &PubSubMessage{
			Tenant: b.tenant,
			Topic:  b.topic,
			SeqID:  i,
		}
	}
	return msgs
}

// PubSubMessage は PubSub service の wire 形式
type PubSubMessage struct {
	Tenant string
	Topic  string
	SeqID  int
}

// 採用初期で実装する builder の placeholder type ----------------------

type workflowMockBuilder struct{}
type decisionMockBuilder struct{}
type secretMockBuilder struct{}
type unimplementedMockBuilder struct{}

File: src/sdk/dotnet/src/K1s0.Sdk.TestFixtures/MockBuilder.cs
// src/sdk/dotnet/src/K1s0.Sdk.TestFixtures/MockBuilder.cs
//
// k1s0 .NET SDK test-fixtures: Mock builder fluent API（領域 3、ADR-TEST-010 §3）。
// リリース時点で 3 service（State / Audit / PubSub）を提供。
// 採用初期で +3 (Workflow / Decision / Secret)、運用拡大時で残 6 を追加。

namespace K1s0.Sdk.TestFixtures;

// 12 service の mock builder への entry point
public sealed class MockBuilderRoot
{
    // 既定 tenant（builder の WithTenant 未指定時に使う）
    private readonly string _defaultTenant;

    internal MockBuilderRoot(string defaultTenant)
    {
        _defaultTenant = defaultTenant;
    }

    // State service の mock builder
    public StateMockBuilder State() => new(_defaultTenant);

    // Audit service の mock builder
    public AuditMockBuilder Audit() => new(_defaultTenant);

    // PubSub service の mock builder
    public PubSubMockBuilder PubSub() => new(_defaultTenant);

    // Workflow / Decision / Secret は採用初期で real 実装する phase marker（標準 .NET 未実装例外ではなく ADR cite 付き独自型を投げる、SHIP_STATUS §9）
    public WorkflowMockBuilder Workflow()
        => throw new FixturePhaseUnsupportedException("Workflow", "採用初期");
    public DecisionMockBuilder Decision()
        => throw new FixturePhaseUnsupportedException("Decision", "採用初期");
    public SecretMockBuilder Secret()
        => throw new FixturePhaseUnsupportedException("Secret", "採用初期");
}

// FixturePhaseUnsupportedException は ADR-TEST-010 §3 段階展開で「採用初期 / 運用拡大時で
// real 実装する」と確定済の builder を呼んだ時に投げる phase marker exception。
// .NET 標準の汎用未実装例外は「未着手」「放置」を示唆するため意図的に避け、
// ADR 名 + 実装 phase を message に含める独自型で「設計上の段階展開」であることを明示する。
public sealed class FixturePhaseUnsupportedException : InvalidOperationException
{
    public string Service { get; }
    public string Phase { get; }
    public FixturePhaseUnsupportedException(string service, string phase)
        : base($"ADR-TEST-010 PHASE: {service} mock builder は{phase}で実装（リリース時点 phase marker）")
    {
        Service = service;
        Phase = phase;
    }
}

// State service mock data の fluent builder
public sealed class StateMockBuilder
{
    private string _tenant;
    private string _key = string.Empty;
    private byte[] _value = Array.Empty<byte>();
    private int _ttl = 0;

    internal StateMockBuilder(string tenant) => _tenant = tenant;

    public StateMockBuilder WithTenant(string tenant) { _tenant = tenant; return this; }
    public StateMockBuilder WithKey(string key) { _key = key; return this; }
    public StateMockBuilder WithValue(byte[] value) { _value = value; return this; }
    public StateMockBuilder WithTTL(int seconds) { _ttl = seconds; return this; }

    // build: 最終的な StateEntry を返す（採用初期で contracts/proto 型に置換）
    public StateEntry Build() => new()
    {
        Tenant = _tenant,
        Key = _key,
        Value = _value,
        TTL = _ttl,
    };
}

// State service の wire 形式
public sealed record StateEntry
{
    public string Tenant { get; init; } = string.Empty;
    public string Key { get; init; } = string.Empty;
    public byte[] Value { get; init; } = Array.Empty<byte>();
    public int TTL { get; init; }
}

// Audit service mock data の fluent builder
public sealed class AuditMockBuilder
{
    private string _tenant;
    private int _entryCount = 0;
    private long _startSeq = 0;

    internal AuditMockBuilder(string tenant) => _tenant = tenant;

    public AuditMockBuilder WithTenant(string tenant) { _tenant = tenant; return this; }
    public AuditMockBuilder WithEntries(int n) { _entryCount = n; return this; }
    public AuditMockBuilder WithSequence(long seq) { _startSeq = seq; return this; }

    public IReadOnlyList<AuditEntry> Build()
    {
        var entries = new List<AuditEntry>(_entryCount);
        for (int i = 0; i < _entryCount; i++)
        {
            entries.Add(new AuditEntry
            {
                Tenant = _tenant,
                Sequence = _startSeq + i,
                // 採用初期で SHA-256 prev_id chain を計算
                PrevID = string.Empty,
            });
        }
        return entries;
    }
}

public sealed record AuditEntry
{
    public string Tenant { get; init; } = string.Empty;
    public long Sequence { get; init; }
    public string PrevID { get; init; } = string.Empty;
}

// PubSub service mock data の fluent builder
public sealed class PubSubMockBuilder
{
    private string _tenant;
    private string _topic = string.Empty;
    private int _messages = 0;
    private int _delayMs = 0;

    internal PubSubMockBuilder(string tenant) => _tenant = tenant;

    public PubSubMockBuilder WithTenant(string tenant) { _tenant = tenant; return this; }
    public PubSubMockBuilder WithTopic(string topic) { _topic = topic; return this; }
    public PubSubMockBuilder WithMessages(int n) { _messages = n; return this; }
    public PubSubMockBuilder WithDelayMs(int ms) { _delayMs = ms; return this; }

    public IReadOnlyList<PubSubMessage> Build()
    {
        var msgs = new List<PubSubMessage>(_messages);
        for (int i = 0; i < _messages; i++)
        {
            msgs.Add(new PubSubMessage
            {
                Tenant = _tenant,
                Topic = _topic,
                SeqID = i,
            });
        }
        return msgs;
    }
}

public sealed record PubSubMessage
{
    public string Tenant { get; init; } = string.Empty;
    public string Topic { get; init; } = string.Empty;
    public long SeqID { get; init; }
}

// 採用初期で実装する builder の placeholder type
public sealed class WorkflowMockBuilder { }
public sealed class DecisionMockBuilder { }
public sealed class SecretMockBuilder { }
